-1

Tested at: b3385da918415e85bc44bb2454790dd60b225632

CMSSW: CMSSW_11_2_X_2020-10-21-2300
SCRAM_ARCH: slc7_amd64_gcc820

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - #31879 @ianna: [DD4hep] Update to master 21 10 2020
 - #31873 @bsunanda: Run3-alca174 Utilize ESGetToken in most of the classes in the package HcalCalibAlgos

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2c34e8/10195/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2c34e8/10195/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2c34e8/10195/summary.html

I found follow errors while testing this PR

**Failed tests:  Build  HeaderConsistency  ClangBuild**

* **Build**:

I found compilation error when building: 

<pre>Copying tmp/slc7_amd64_gcc820/src/Alignment/CocoaApplication/bin/cocoa/cocoa to productstore area:
Entering library rule at Alignment/CocoaApplication
>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-21-2300/src/Alignment/CocoaApplication/src/CocoaAnalyzer.cc
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-21-2300/src/Alignment/CocoaApplication/src/CocoaAnalyzer.cc:6:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-21-2300/src/DetectorDescription/DDCMS/interface/DDCompactView.h: In function 'std::vector<T> cms::getAllParameterValuesFromSpecParSections(const DDSpecParRegistry&, const string&, const string&)':
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-21-2300/src/DetectorDescription/DDCMS/interface/DDCompactView.h:61:28: error: request for member 'second' in 'mySpecParSection', which is of pointer type 'const dd4hep::SpecPar* const' (maybe you meant to use '->' ?)
       if (mySpecParSection.second->hasPath(nodePath)) {
                            ^~~~~~
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-21-2300/src/DetectorDescription/DDCMS/interface/DDCompactView.h:62:33: error: request for member 'second' in 'mySpecParSection', which is of pointer type 'const dd4hep::SpecPar* const' (maybe you meant to use '->' ?)
         return mySpecParSection.second->value<std::vector<T>>(parameterName);
                                 ^~~~~~
</pre>

* **Clang**:

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```


<pre>>> Entering Package Alignment/CocoaApplication
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-21-2300/src/Alignment/CocoaApplication/bin/cocoa.cpp
Entering library rule at Alignment/CocoaApplication
>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-21-2300/src/Alignment/CocoaApplication/src/CocoaAnalyzer.cc
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-21-2300/src/Alignment/CocoaApplication/src/CocoaAnalyzer.cc:6:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-21-2300/src/DetectorDescription/DDCMS/interface/DDCompactView.h:61:27: error: member reference type 'const dd4hep::SpecPar *const' is a pointer; did you mean to use '->'?
      if (mySpecParSection.second->hasPath(nodePath)) {
          ~~~~~~~~~~~~~~~~^
                          ->
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-21-2300/src/DetectorDescription/DDCMS/interface/DDCompactView.h:61:28: error: no member named 'second' in 'dd4hep::SpecPar'
      if (mySpecParSection.second->hasPath(nodePath)) {
</pre>


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - #31879 @ianna: [DD4hep] Update to master 21 10 2020
 - #31873 @bsunanda: Run3-alca174 Utilize ESGetToken in most of the classes in the package HcalCalibAlgos

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2c34e8/10195/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2c34e8/10195/git-merge-result

--- a/DetectorDescription/DDCMS/interface/DDNamespace.h
+++ b/DetectorDescription/DDCMS/interface/DDNamespace.h
@@ -70,6 +70,7 @@ namespace cms {
     DDParsingContext* setContext() { return m_context; }
 
     std::string_view name() const { return m_name; }
+    std::string noNamespace(const std::string&) const;
 
     std::vector<double> vecDbl(const std::string& name) const;
     std::vector<float> vecFloat(const std::string& name) const;

--- a/DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc
+++ b/DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc
@@ -758,7 +758,7 @@ void Converter<DDLRotationByAxis>::operator()(xml_h element) const {
   xml_dim_t xrot(element);
   xml_dim_t par(xrot.parent());
   if (xrot.hasAttr(_U(name))) {
-    string nam = xrot.nameStr();  // + string("Rotation"); // xrot.hasAttr(_U(name)) ? xrot.nameStr() : par.nameStr();
+    string nam = xrot.nameStr();
     string axis = ns.attr<string>(xrot, DD_CMU(axis));
     double angle = ns.attr<double>(xrot, _U(angle));
     Rotation3D rot;
@@ -786,7 +786,7 @@ void Converter<DDLLogicalPart>::operator()(xml_h element) const {
   xml_dim_t e(element);
   string sol = e.child(DD_CMU(rSolid)).attr<string>(_U(name));
   string mat = e.child(DD_CMU(rMaterial)).attr<string>(_U(name));
-  string volName = e.attr<string>(_U(name));
+  string volName = ns.prepend(e.attr<string>(_U(name)));
   Solid solid = ns.solid(sol);
   Material material = ns.material(mat);
 
@@ -852,8 +852,8 @@ void Converter<DDLPosPart>::operator()(xml_h element) const {
   cms::DDNamespace ns(_param<cms::DDParsingContext>());  //, element, true );
   xml_dim_t e(element);
   int copy = e.attr<int>(DD_CMU(copyNumber));
-  string parentName = ns.attr<string>(e.child(DD_CMU(rParent)), _U(name));
-  string childName = ns.attr<string>(e.child(DD_CMU(rChild)), _U(name));
+  string parentName = ns.prepend(ns.attr<string>(e.child(DD_CMU(rParent)), _U(name)));
+  string childName = ns.prepend(ns.attr<string>(e.child(DD_CMU(rChild)), _U(name)));
   Volume parent = ns.volume(parentName, false);
   Volume child = ns.volume(childName, false);
 

--- a/DetectorDescription/DDCMS/src/DDNamespace.cc
+++ b/DetectorDescription/DDCMS/src/DDNamespace.cc
@@ -143,7 +143,7 @@ const dd4hep::Rotation3D& DDNamespace::rotation(const string& name) const {
 }
 
 dd4hep::Volume DDNamespace::addVolumeNS(dd4hep::Volume vol) const {
-  string n = vol.name();
+  string n = prepend(vol.name());
   dd4hep::Solid s = vol.solid();
   dd4hep::Material m = vol.material();
   vol->SetName(n.c_str());
@@ -277,4 +277,14 @@ std::vector<float> DDNamespace::vecFloat(const std::string& name) const {
     return result;
   } else
     return std::vector<float>();
+}
+
+std::string DDNamespace::noNamespace(const std::string& fullName) const {
+  std::string result(fullName);
+  auto n = result.find(':');
+  if (n == std::string::npos) {
+    return result;
+  } else {
+    return result.substr(n + 1);
+  }
 }

--- a/DetectorDescription/DDCMS/test/DDFilteredView.find.cppunit.cc
+++ b/DetectorDescription/DDCMS/test/DDFilteredView.find.cppunit.cc
@@ -92,7 +92,7 @@ void testDDFilteredViewFind::checkFilteredView() {
   fview.goTo({0, 0, 8});
   printMe(fview);
 
-  CPPUNIT_ASSERT(fview.name() == "MUON");
+  CPPUNIT_ASSERT(fview.name() == "muonBase:MUON");
 
   // Go to the first daughter
   fview.next(0);
@@ -129,7 +129,7 @@ void testDDFilteredViewFind::checkFilteredView() {
   std::cout << "\n==== Let's do it again, go to Muon\n";
   fview.goTo({0, 0, 8});
   printMe(fview);
-  CPPUNIT_ASSERT(fview.name() == "MUON");
+  CPPUNIT_ASSERT(fview.name() == "muonBase:MUON");
 
   // Go to the first daughter
   fview.next(0);

--- a/DetectorDescription/DDCMS/test/DDFilteredView.get.cppunit.cc
+++ b/DetectorDescription/DDCMS/test/DDFilteredView.get.cppunit.cc
@@ -95,7 +95,7 @@ void testDDFilteredViewGet::checkFilteredView() {
   fview.goTo({0, 0, 8});
   printMe(fview);
 
-  CPPUNIT_ASSERT(fview.name() == "MUON");
+  CPPUNIT_ASSERT(fview.name() == "muonBase:MUON");
 
   // Go to the first daughter
   fview.next(0);
@@ -132,7 +132,7 @@ void testDDFilteredViewGet::checkFilteredView() {
   std::cout << "\n==== Let's do it again, go to Muon\n";
   fview.goTo({0, 0, 8});
   printMe(fview);
-  CPPUNIT_ASSERT(fview.name() == "MUON");
+  CPPUNIT_ASSERT(fview.name() == "muonBase:MUON");
 
   // Go to the first daughter
   fview.next(0);

--- a/DetectorDescription/DDCMS/test/DDFilteredView.goto.cppunit.cc
+++ b/DetectorDescription/DDCMS/test/DDFilteredView.goto.cppunit.cc
@@ -86,7 +86,7 @@ void testDDFilteredViewGoTo::checkFilteredView() {
   // Start with Tracker
   std::cout << "\n==== Let's go to Tracker\n";
   fview.goTo({0, 0, 6});
-  CPPUNIT_ASSERT(fview.name() == "Tracker");
+  CPPUNIT_ASSERT(fview.name() == "tracker:Tracker");
   printMe(fview);
 
   // Go to the first daughter
@@ -123,7 +123,7 @@ void testDDFilteredViewGoTo::checkFilteredView() {
 
   std::cout << "\n==== Let's do it again, go to Tracker\n";
   fview.goTo({0, 0, 6});
-  CPPUNIT_ASSERT(fview.name() == "Tracker");
+  CPPUNIT_ASSERT(fview.name() == "tracker:Tracker");
   printMe(fview);
 
   // Go to the first daughter

--- a/Geometry/MTDCommonData/test/DD4hep_TestMTDIdealGeometry.cc
+++ b/Geometry/MTDCommonData/test/DD4hep_TestMTDIdealGeometry.cc
@@ -141,10 +141,10 @@ void DD4hep_TestMTDIdealGeometry::analyze(const edm::Event& iEvent, const edm::E
   uint32_t level(0);
 
   do {
-    if (fv.name() == "BarrelTimingLayer") {
+    if (dd4hep::dd::noNamespace(fv.name()) == "BarrelTimingLayer") {
       isBarrel = true;
       edm::LogInfo("DD4hep_TestMTDIdealGeometry") << "isBarrel = " << isBarrel;
-    } else if (fv.name() == "EndcapTimingLayer") {
+    } else if (dd4hep::dd::noNamespace(fv.name()) == "EndcapTimingLayer") {
       isBarrel = false;
       edm::LogInfo("DD4hep_TestMTDIdealGeometry") << "isBarrel = " << isBarrel;
     }
@@ -168,7 +168,7 @@ void DD4hep_TestMTDIdealGeometry::analyze(const edm::Event& iEvent, const edm::E
       write = false;
       exitLoop = true;
     }
-    if (fv.name() == ddTopNodeName_) {
+    if (dd4hep::dd::noNamespace(fv.name()) == ddTopNodeName_) {
       write = true;
       level = fv.navPos().size();
     }
@@ -194,7 +194,7 @@ void DD4hep_TestMTDIdealGeometry::analyze(const edm::Event& iEvent, const edm::E
 
       for (auto const& t : specs) {
         for (auto const& it : t.second->paths) {
-          if (dd4hep::dd::compareEqual(fv.name(), dd4hep::dd::realTopName(it))) {
+          if (dd4hep::dd::compareEqual(dd4hep::dd::noNamespace(fv.name()), dd4hep::dd::realTopName(it))) {
             isSens = true;
             break;
           }

--- a/Geometry/TrackerCommonData/plugins/dd4hep/DDTECModuleAlgo.cc
+++ b/Geometry/TrackerCommonData/plugins/dd4hep/DDTECModuleAlgo.cc
@@ -164,7 +164,7 @@ static long algorithm(Detector& /* description */, cms::DDParsingContext& ctxt, 
     tag = "Stereo";
   //usefull constants
   const double topFrameEndZ = 0.5 * (-waferPosition + fullHeight) + pitchHeight + hybridHeight - topFrameHeight;
-  string idName = ns.prepend(ns.realName(mother.name()));
+  string idName = ns.noNamespace(mother.name());
   edm::LogVerbatim("TECGeom") << "idName: " << idName << " parent " << mother.name() << " namespace " << ns.name();
   Solid solid;
 

--- a/Geometry/TrackerCommonData/plugins/dd4hep/DDTrackerZPosAlgo.cc
+++ b/Geometry/TrackerCommonData/plugins/dd4hep/DDTrackerZPosAlgo.cc
@@ -1,7 +1,6 @@
 #include "DD4hep/DetFactoryHelper.h"
 #include "DetectorDescription/DDCMS/interface/DDPlugins.h"
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
-#include "FWCore/MessageLogger/interface/MessageLogger.h"
 
 using namespace std;
 using namespace dd4hep;

